service/matching/taskQueueManager.go

```diff
@@ -285,13 +311,59 @@ func (c *taskQueueManagerImpl) Start() {
 		return
 	}
 	c.liveness.Start()
-	c.taskWriter.Start()
-	c.taskReader.Start()
+	c.StartDbTaskManagement()
+	c.dbTaskManagementUpdatedAt = time.Now()
 	go c.fetchMetadataFromRootPartitionOnInit(context.TODO())
 	c.logger.Info("", tag.LifeCycleStarted)
 	c.metricScope.IncCounter(metrics.TaskQueueStartedCounter)
 }
 
+func (c *taskQueueManagerImpl) StartDbTaskManagement() {
+	c.enableDbTaskManager = c.config.EnableDbTaskManager()
+	if c.enableDbTaskManager {
+		c.dbTaskManager.Start()
+	} else {
+		// TODO(deprecate)
+		c.taskWriter.Start()
+		c.taskReader.Start()
+		c.initializedError.Get(context.TODO())
+	}
+}
+
+func (c *taskQueueManagerImpl) SwitchDbTaskManagementIfNecessary() bool {
+	c.dbTaskManagementLock.Lock()
+	defer c.dbTaskManagementLock.Unlock()
+	ts := time.Now()
+	if ts.Sub(c.dbTaskManagementUpdatedAt) < dbTaskManagementSwitchThreshold {
+		return c.enableDbTaskManager
+	}
+	c.dbTaskManagementUpdatedAt = ts
+	newVal := c.config.EnableDbTaskManager()
+	if c.enableDbTaskManager == newVal {
+		return c.enableDbTaskManager
+	}
+	c.StopDbTaskManagement()
+	c.StartDbTaskManagement()
```

Comment:
I don't think `EnableDbTaskManager()` can be evaluated multiple times here. What if the value change in the middle of the two function call.